crates/shielded_token/src/masp/shielded_sync/dispatcher.rs

```diff
@@ -381,7 +382,101 @@ where
         }
     }
 
-    fn apply_cache_to_shielded_context(
+    // Search for the Transaction order usedby the protocol using trial and
+    // error.
+    async fn transaction_order_search(
+        &mut self,
+        last_witnessed_tx: &Option<IndexedTx>,
+    ) -> Result<Vec<(IndexedTx, Transaction)>, eyre::Error> {
+        let mut ordered_txs = vec![];
+        // Get the unordered Transactions with their indices
+        let mut indexed_txs = self.cache.fetched.take().into_iter().peekable();
+        let needs_witness_map_update =
+            self.client.capabilities().needs_witness_map_update();
+        while let Some(first_tx) = indexed_txs.next() {
+            // Take only Transactions belonging to one block
+            let mut block_txs = vec![first_tx];
+            while let Some(next_tx) = indexed_txs
```

Comment:
This can be simplified using `take_while` and `collect`. Then we don't need the iterator to be peekable. You might need to combine the `take_while` with the iterator take `by_ref`. Not sure

Reply:
Well shucks.